# Extracción, Transformación y Carga de datos (ETL)

### Importamos librerías

In [33]:
import pandas as pd
import numpy as np
import json
import ast
import warnings
warnings.filterwarnings("ignore")
import sys
sys.path.insert(0, '../')
import Herramientas as Herr

### Carga de datos

In [34]:
filas = []

with open('../datasets/australian_user_reviews.json',encoding='utf-8') as r:
    for lines in r.readlines():
        filas.append(ast.literal_eval(lines))
        

data_reviews = pd.DataFrame(filas)
data_reviews

user_id                                           user_url  \
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1                js41637               http://steamcommunity.com/id/js41637   
2              evcentric             http://steamcommunity.com/id/evcentric   
3                  doctr                 http://steamcommunity.com/id/doctr   
4              maplemage             http://steamcommunity.com/id/maplemage   
...                  ...                                                ...   
25794  76561198306599751  http://steamcommunity.com/profiles/76561198306...   
25795           Ghoustik              http://steamcommunity.com/id/Ghoustik   
25796  76561198310819422  http://steamcommunity.com/profiles/76561198310...   
25797  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
25798        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   

                                                 reviews  
0      [{'funny': '', 'posted': 'Posted November 5, 2...  
1      [{'funny': '', 'posted': 'Posted June 24, 2014...  
2      [{'funny': '', 'posted': 'Posted February 3.',...  
3      [{'funny': '', 'posted': 'Posted October 14, 2...  
4      [{'funny': '3 people found this review funny',...  
...                                                  ...  
25794  [{'funny': '', 'posted': 'Posted May 31.', 'la...  
25795  [{'funny': '', 'posted': 'Posted June 17.', 'l...  
25796  [{'funny': '1 person found this review funny',...  
25797  [{'funny': '', 'posted': 'Posted July 21.', 'l...  
25798  [{'funny': '1 person found this review funny',...  

[25799 rows x 3 columns]

In [35]:
Herr.analizar_datos(data_reviews)

Nombre Tipos de Datos Únicos  % de Valores No Nulos  % de Valores Nulos  \
0   user_id       [<class 'str'>]                  100.0                 0.0   
1  user_url       [<class 'str'>]                  100.0                 0.0   
2   reviews      [<class 'list'>]                  100.0                 0.0   

   Cantidad de Valores Nulos  
0                          0  
1                          0  
2                          0

### Transformación de los datos

In [36]:
# Se verifica si hay duplicados de filas y numero de fila 
duplicados = data_reviews['user_id'].duplicated()
print(duplicados.value_counts())
print(110*'-')
filas_duplicadas = data_reviews.loc[duplicados,:]
filas_duplicadas

user_id
False    25485
True       314
Name: count, dtype: int64
--------------------------------------------------------------------------------------------------------------


user_id                                           user_url  \
456           bokkkbokkk            http://steamcommunity.com/id/bokkkbokkk   
1182          ImSeriouss            http://steamcommunity.com/id/ImSeriouss   
1456   76561198062039159  http://steamcommunity.com/profiles/76561198062...   
1477   76561198045009232  http://steamcommunity.com/profiles/76561198045...   
1746        nitr0ticwolf          http://steamcommunity.com/id/nitr0ticwolf   
...                  ...                                                ...   
17819  76561198076474887  http://steamcommunity.com/profiles/76561198076...   
17916        yolofaceguy           http://steamcommunity.com/id/yolofaceguy   
18028  76561198075591109  http://steamcommunity.com/profiles/76561198075...   
18234  76561198092022514  http://steamcommunity.com/profiles/76561198092...   
18309     SuchGayMuchWow        http://steamcommunity.com/id/SuchGayMuchWow   

                                                 reviews  
456    [{'funny': '', 'posted': 'Posted September 24,...  
1182   [{'funny': '', 'posted': 'Posted January 10, 2...  
1456   [{'funny': '', 'posted': 'Posted August 24, 20...  
1477   [{'funny': '', 'posted': 'Posted October 31, 2...  
1746   [{'funny': '', 'posted': 'Posted December 12, ...  
...                                                  ...  
17819  [{'funny': '', 'posted': 'Posted April 12.', '...  
17916  [{'funny': '', 'posted': 'Posted October 31, 2...  
18028  [{'funny': '', 'posted': 'Posted December 26, ...  
18234  [{'funny': '', 'posted': 'Posted July 3.', 'la...  
18309  [{'funny': '', 'posted': 'Posted July 18, 2015...  

[314 rows x 3 columns]

Se intento con varios id diferentes de los duplicados y encontramos que ninguno de ellos repite los comentarios, significa que el mismo usuario a hecho mas de un comentario 

In [37]:
id = '76561198062039159'
resenia = filas_duplicadas[filas_duplicadas['user_id'] == id]['reviews']

for x in resenia:   
    for r in x:
        print(r['review'])
    print(100*'--')

Doto of the ancients > LoLegend league of Legends10/IV would play again, best fps in today's scientific community
This is possibly the best tactical FPS i've ever played.Insurgency provides hours of fun and it also looks fabutabulous. It has amazing gunplay and guns, fantastic supression mechanics and somewhat good people.This game is a 10/10, no questions asked.
Buy it and feel bad when a stupid ♥♥♥♥ing robot kills your dudes.It's really good.
I rate it R8/Revolver
A great game with excellent gameplay, hours of fun and a good turn based system of doing practically everything.A must have if you've got hours to spare.
Buy it, i don't care what you have to say, just buy it.It's worth the three dollars so just ♥♥♥♥ING BUY IT.
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------


In [38]:
data_reviews2= pd.json_normalize(data_reviews['reviews'])
data_reviews2


0  \
0      {'funny': '', 'posted': 'Posted November 5, 20...   
1      {'funny': '', 'posted': 'Posted June 24, 2014....   
2      {'funny': '', 'posted': 'Posted February 3.', ...   
3      {'funny': '', 'posted': 'Posted October 14, 20...   
4      {'funny': '3 people found this review funny', ...   
...                                                  ...   
25794  {'funny': '', 'posted': 'Posted May 31.', 'las...   
25795  {'funny': '', 'posted': 'Posted June 17.', 'la...   
25796  {'funny': '1 person found this review funny', ...   
25797  {'funny': '', 'posted': 'Posted July 21.', 'la...   
25798  {'funny': '1 person found this review funny', ...   

                                                       1  \
0      {'funny': '', 'posted': 'Posted July 15, 2011....   
1      {'funny': '', 'posted': 'Posted September 8, 2...   
2      {'funny': '', 'posted': 'Posted December 4, 20...   
3      {'funny': '', 'posted': 'Posted July 28, 2012....   
4      {'funny': '1 person found this review funny', ...   
...                                                  ...   
25794                                               None   
25795                                               None   
25796                                               None   
25797  {'funny': '', 'posted': 'Posted July 10.', 'la...   
25798  {'funny': '', 'posted': 'Posted July 20.', 'la...   

                                                       2  \
0      {'funny': '', 'posted': 'Posted April 21, 2011...   
1      {'funny': '', 'posted': 'Posted November 29, 2...   
2      {'funny': '', 'posted': 'Posted November 3, 20...   
3      {'funny': '', 'posted': 'Posted June 2, 2012.'...   
4      {'funny': '2 people found this review funny', ...   
...                                                  ...   
25794                                               None   
25795                                               None   
25796                                               None   
25797  {'funny': '', 'posted': 'Posted July 10.', 'la...   
25798  {'funny': '', 'posted': 'Posted July 2.', 'las...   

                                                       3  \
0                                                   None   
1                                                   None   
2      {'funny': '', 'posted': 'Posted October 15, 20...   
3      {'funny': '', 'posted': 'Posted June 29, 2014....   
4      {'funny': '', 'posted': 'Posted July 11, 2013....   
...                                                  ...   
25794                                               None   
25795                                               None   
25796                                               None   
25797  {'funny': '', 'posted': 'Posted July 8.', 'las...   
25798                                               None   

                                                       4  \
0                                                   None   
1                                                   None   
2      {'funny': '', 'posted': 'Posted October 15, 20...   
3      {'funny': '', 'posted': 'Posted November 22, 2...   
4                                                   None   
...                                                  ...   
25794                                               None   
25795                                               None   
25796                                               None   
25797                                               None   
25798                                               None   

                                                       5     6     7     8  \
0                                                   None  None  None  None   
1                                                   None  None  None  None   
2      {'funny': '', 'posted': 'Posted October 15, 20...  None  None  None   
3      {'funny': '', 'posted': 'Posted February 23, 2...  None  None  None   
4                                                   None  No

In [39]:
data_reviews3 = pd.concat([data_reviews,data_reviews2],axis=1)
data_reviews3

user_id                                           user_url  \
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1                js41637               http://steamcommunity.com/id/js41637   
2              evcentric             http://steamcommunity.com/id/evcentric   
3                  doctr                 http://steamcommunity.com/id/doctr   
4              maplemage             http://steamcommunity.com/id/maplemage   
...                  ...                                                ...   
25794  76561198306599751  http://steamcommunity.com/profiles/76561198306...   
25795           Ghoustik              http://steamcommunity.com/id/Ghoustik   
25796  76561198310819422  http://steamcommunity.com/profiles/76561198310...   
25797  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
25798        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   

                                                 reviews  \
0      [{'funny': '', 'posted': 'Posted November 5, 2...   
1      [{'funny': '', 'posted': 'Posted June 24, 2014...   
2      [{'funny': '', 'posted': 'Posted February 3.',...   
3      [{'funny': '', 'posted': 'Posted October 14, 2...   
4      [{'funny': '3 people found this review funny',...   
...                                                  ...   
25794  [{'funny': '', 'posted': 'Posted May 31.', 'la...   
25795  [{'funny': '', 'posted': 'Posted June 17.', 'l...   
25796  [{'funny': '1 person found this review funny',...   
25797  [{'funny': '', 'posted': 'Posted July 21.', 'l...   
25798  [{'funny': '1 person found this review funny',...   

                                                       0  \
0      {'funny': '', 'posted': 'Posted November 5, 20...   
1      {'funny': '', 'posted': 'Posted June 24, 2014....   
2      {'funny': '', 'posted': 'Posted February 3.', ...   
3      {'funny': '', 'posted': 'Posted October 14, 20...   
4      {'funny': '3 people found this review funny', ...   
...                                                  ...   
25794  {'funny': '', 'posted': 'Posted May 31.', 'las...   
25795  {'funny': '', 'posted': 'Posted June 17.', 'la...   
25796  {'funny': '1 person found this review funny', ...   
25797  {'funny': '', 'posted': 'Posted July 21.', 'la...   
25798  {'funny': '1 person found this review funny', ...   

                                                       1  \
0      {'funny': '', 'posted': 'Posted July 15, 2011....   
1      {'funny': '', 'posted': 'Posted September 8, 2...   
2      {'funny': '', 'posted': 'Posted December 4, 20...   
3      {'funny': '', 'posted': 'Posted July 28, 2012....   
4      {'funny': '1 person found this review funny', ...   
...                                                  ...   
25794                                               None   
25795                                               None   
25796                                               None   
25797  {'funny': '', 'posted': 'Posted July 10.', 'la...   
25798  {'funny': '', 'posted': 'Posted July 20.', 'la...   

                                                       2  \
0      {'funny': '', 'posted': 'Posted April 21, 2011...   
1      {'funny': '', 'posted': 'Posted November 29, 2...   
2      {'funny': '', 'posted': 'Posted November 3, 20...   
3      {'funny': '', 'posted': 'Posted June 2, 2012.'...   
4      {'funny': '2 people found this review funny', ...   
...                                                  ...   
25794                                               None   
25795                                               None   
25796                                               None   
25797  {'funny': '', 'posted': 'Posted July 10.', 'la...   
25798  {'funny': '', 'posted': 'Posted July 2.', 'las...   

                                                       3  \
0                                                   None   
1                                                   None   
2   

In [40]:
data_reviews3 = data_reviews3.drop('reviews',axis=1)

In [41]:
data_reviews3 = pd.melt(data_reviews3,id_vars=['user_id','user_url'],value_vars=list(range(9)),value_name='reviews')
data_reviews

user_id                                           user_url  \
0      76561197970982479  http://steamcommunity.com/profiles/76561197970...   
1                js41637               http://steamcommunity.com/id/js41637   
2              evcentric             http://steamcommunity.com/id/evcentric   
3                  doctr                 http://steamcommunity.com/id/doctr   
4              maplemage             http://steamcommunity.com/id/maplemage   
...                  ...                                                ...   
25794  76561198306599751  http://steamcommunity.com/profiles/76561198306...   
25795           Ghoustik              http://steamcommunity.com/id/Ghoustik   
25796  76561198310819422  http://steamcommunity.com/profiles/76561198310...   
25797  76561198312638244  http://steamcommunity.com/profiles/76561198312...   
25798        LydiaMorley           http://steamcommunity.com/id/LydiaMorley   

                                                 reviews  
0      [{'funny': '', 'posted': 'Posted November 5, 2...  
1      [{'funny': '', 'posted': 'Posted June 24, 2014...  
2      [{'funny': '', 'posted': 'Posted February 3.',...  
3      [{'funny': '', 'posted': 'Posted October 14, 2...  
4      [{'funny': '3 people found this review funny',...  
...                                                  ...  
25794  [{'funny': '', 'posted': 'Posted May 31.', 'la...  
25795  [{'funny': '', 'posted': 'Posted June 17.', 'l...  
25796  [{'funny': '1 person found this review funny',...  
25797  [{'funny': '', 'posted': 'Posted July 21.', 'l...  
25798  [{'funny': '1 person found this review funny',...  

[25799 rows x 3 columns]

In [42]:
Herr.analizar_datos(data_reviews3)

Nombre                 Tipos de Datos Únicos  % de Valores No Nulos  \
0   user_id                       [<class 'str'>]                 100.00   
1  user_url                       [<class 'str'>]                 100.00   
2  variable                       [<class 'int'>]                 100.00   
3   reviews  [<class 'dict'>, <class 'NoneType'>]                  25.43   

   % de Valores Nulos  Cantidad de Valores Nulos  
0                0.00                          0  
1                0.00                          0  
2                0.00                          0  
3               74.57                     173155

In [43]:
user = 'LydiaMorley'
filtrado = data_reviews3[data_reviews3['user_id'] == user]
filtrado

user_id                                  user_url variable  \
25798   LydiaMorley  http://steamcommunity.com/id/LydiaMorley        0   
51597   LydiaMorley  http://steamcommunity.com/id/LydiaMorley        1   
77396   LydiaMorley  http://steamcommunity.com/id/LydiaMorley        2   
103195  LydiaMorley  http://steamcommunity.com/id/LydiaMorley        3   
128994  LydiaMorley  http://steamcommunity.com/id/LydiaMorley        4   
154793  LydiaMorley  http://steamcommunity.com/id/LydiaMorley        5   
180592  LydiaMorley  http://steamcommunity.com/id/LydiaMorley        6   
206391  LydiaMorley  http://steamcommunity.com/id/LydiaMorley        7   
232190  LydiaMorley  http://steamcommunity.com/id/LydiaMorley        8   

                                                  reviews  
25798   {'funny': '1 person found this review funny', ...  
51597   {'funny': '', 'posted': 'Posted July 20.', 'la...  
77396   {'funny': '', 'posted': 'Posted July 2.', 'las...  
103195                                               None  
128994                                               None  
154793                                               None  
180592                                               None  
206391                                               None  
232190                                               None

In [44]:
data_reviews3 = data_reviews3.dropna()
Herr.analizar_datos(data_reviews3)

Nombre Tipos de Datos Únicos  % de Valores No Nulos  % de Valores Nulos  \
0   user_id       [<class 'str'>]                  100.0                 0.0   
1  user_url       [<class 'str'>]                  100.0                 0.0   
2  variable       [<class 'int'>]                  100.0                 0.0   
3   reviews      [<class 'dict'>]                  100.0                 0.0   

   Cantidad de Valores Nulos  
0                          0  
1                          0  
2                          0  
3                          0

In [45]:
data_des = data_reviews3['reviews'].apply(pd.Series)
data_des


funny                     posted  \
0                                          Posted November 5, 2011.   
1                                             Posted June 24, 2014.   
2                                                Posted February 3.   
3                                          Posted October 14, 2013.   
4       3 people found this review funny     Posted April 15, 2014.   
...                                  ...                        ...   
231919                                      Posted August 15, 2014.   
231921                                       Posted August 2, 2014.   
232047                                        Posted July 31, 2015.   
232127                                    Posted December 20, 2015.   
232129                                           Posted February 3.   

                          last_edited item_id  \
0                                        1250   
1                                      251610   
2                                      248820   
3                                      250320   
4                                      211420   
...                               ...     ...   
231919  Last edited November 3, 2014.     440   
231921                                 304930   
232047                                 265630   
232127                                 304050   
232129                                    570   

                                                helpful  recommend  \
0                                        No ratings yet       True   
1       15 of 20 people (75%) found this review helpful       True   
2                                        No ratings yet       True   
3        2 of 2 people (100%) found this review helpful       True   
4       35 of 43 people (81%) found this review helpful       True   
...                                                 ...        ...   
231919                                   No ratings yet       True   
231921                                   No ratings yet       True   
232047                                   No ratings yet       True   
232127                                   No ratings yet       True   
232129     0 of 2 people (0%) found this review helpful      False   

                                                   review  
0       Simple yet with great replayability. In my opi...  
1       I know what you think when you see this title ...  
2       A suitably punishing roguelike platformer.  Wi...  
3       This game... is so fun. The fight sequences ha...  
4                                                 Git gud  
...                                                   ...  
231919  TF2 is alot of fun and its really good but the...  
231921                              Fun game with friends  
232047                                        So Fun!! :D  
232127  This game is great. The only thing is,Why cant...  
232129  DOTA 2.normal skill - 2ez for me>high skill - ...  

[59036 rows x 7 columns]

In [46]:
data_normalizado = pd.concat([data_reviews3[['user_id','user_url']],data_des],axis=1)
data_normalizado

user_id  \
0           76561197970982479   
1                     js41637   
2                   evcentric   
3                       doctr   
4                   maplemage   
...                       ...   
231919     SKELETRONPRIMEISOP   
231921      76561198141079508   
232047            ShadowYT100   
232127  bestcustomurlevermade   
232129      76561198209894493   

                                                 user_url  \
0       http://steamcommunity.com/profiles/76561197970...   
1                    http://steamcommunity.com/id/js41637   
2                  http://steamcommunity.com/id/evcentric   
3                      http://steamcommunity.com/id/doctr   
4                  http://steamcommunity.com/id/maplemage   
...                                                   ...   
231919    http://steamcommunity.com/id/SKELETRONPRIMEISOP   
231921  http://steamcommunity.com/profiles/76561198141...   
232047           http://steamcommunity.com/id/ShadowYT100   
232127  http://steamcommunity.com/id/bestcustomurlever...   
232129  http://steamcommunity.com/profiles/76561198209...   

                                   funny                     posted  \
0                                          Posted November 5, 2011.   
1                                             Posted June 24, 2014.   
2                                                Posted February 3.   
3                                          Posted October 14, 2013.   
4       3 people found this review funny     Posted April 15, 2014.   
...                                  ...                        ...   
231919                                      Posted August 15, 2014.   
231921                                       Posted August 2, 2014.   
232047                                        Posted July 31, 2015.   
232127                                    Posted December 20, 2015.   
232129                                           Posted February 3.   

                          last_edited item_id  \
0                                        1250   
1                                      251610   
2                                      248820   
3                                      250320   
4                                      211420   
...                               ...     ...   
231919  Last edited November 3, 2014.     440   
231921                                 304930   
232047                                 265630   
232127                                 304050   
232129                                    570   

                                                helpful  recommend  \
0                                        No ratings yet       True   
1       15 of 20 people (75%) found this review helpful       True   
2                                        No ratings yet       True   
3        2 of 2 people (100%) found this review helpful       True   
4       35 of 43 people (81%) found this review helpful       True   
...                                                 ...        ...   
231919                                   No ratings yet       True   
231921                                   No ratings yet       True   
232047                                   No ratings yet       True   
232127                                   No ratings yet       True   
232129     0 of 2 people (0%) found this review helpful      False   

                                                   review  
0       Simple yet with great replayability. In my opi...  
1       I know what you think when you see this title ...  
2       A suitably punishing roguelike platformer.  Wi...  
3       This game... is so fun. The fight sequences ha...  
4                                                 Git gud  
...                                                   ...  
231919  TF2 is alot of fun and its really good but the...  
231921                              Fun game with friends  
232047                                        So Fun!! :D  
232127  T

Ahora tenemos que revisar cuantos nulos o vacios tienen las columnas, hay columnas que en lugar de tener valor None poseen espacios vacios, lo que no da ningun valor que nos sume.

In [47]:
Herr.analizar_datos(data_normalizado)

Nombre Tipos de Datos Únicos  % de Valores No Nulos  \
0      user_id       [<class 'str'>]                  100.0   
1     user_url       [<class 'str'>]                  100.0   
2        funny       [<class 'str'>]                  100.0   
3       posted       [<class 'str'>]                  100.0   
4  last_edited       [<class 'str'>]                  100.0   
5      item_id       [<class 'str'>]                  100.0   
6      helpful       [<class 'str'>]                  100.0   
7    recommend      [<class 'bool'>]                  100.0   
8       review       [<class 'str'>]                  100.0   

   % de Valores Nulos  Cantidad de Valores Nulos  
0                 0.0                          0  
1                 0.0                          0  
2                 0.0                          0  
3                 0.0                          0  
4                 0.0                          0  
5                 0.0                          0  
6                 0.0                          0  
7                 0.0                          0  
8                 0.0                          0

In [48]:
data_normalizado = data_normalizado.replace('', None)
Herr.analizar_datos(data_normalizado)

Nombre                Tipos de Datos Únicos  % de Valores No Nulos  \
0      user_id                      [<class 'str'>]                 100.00   
1     user_url                      [<class 'str'>]                 100.00   
2        funny  [<class 'NoneType'>, <class 'str'>]                  13.77   
3       posted                      [<class 'str'>]                 100.00   
4  last_edited  [<class 'NoneType'>, <class 'str'>]                  10.30   
5      item_id                      [<class 'str'>]                 100.00   
6      helpful                      [<class 'str'>]                 100.00   
7    recommend                     [<class 'bool'>]                 100.00   
8       review  [<class 'str'>, <class 'NoneType'>]                  99.95   

   % de Valores Nulos  Cantidad de Valores Nulos  
0                0.00                          0  
1                0.00                          0  
2               86.23                      50904  
3                0.00                          0  
4               89.70                      52953  
5                0.00                          0  
6                0.00                          0  
7                0.00                          0  
8                0.05                         30

Las columnas funny (86.23%) y last edited(89.7%) tienen un gran porcentaje de nulos, seran eliminadas

In [49]:
data_normalizado = data_normalizado.drop(columns=['funny','last_edited'])
data_normalizado

user_id  \
0           76561197970982479   
1                     js41637   
2                   evcentric   
3                       doctr   
4                   maplemage   
...                       ...   
231919     SKELETRONPRIMEISOP   
231921      76561198141079508   
232047            ShadowYT100   
232127  bestcustomurlevermade   
232129      76561198209894493   

                                                 user_url  \
0       http://steamcommunity.com/profiles/76561197970...   
1                    http://steamcommunity.com/id/js41637   
2                  http://steamcommunity.com/id/evcentric   
3                      http://steamcommunity.com/id/doctr   
4                  http://steamcommunity.com/id/maplemage   
...                                                   ...   
231919    http://steamcommunity.com/id/SKELETRONPRIMEISOP   
231921  http://steamcommunity.com/profiles/76561198141...   
232047           http://steamcommunity.com/id/ShadowYT100   
232127  http://steamcommunity.com/id/bestcustomurlever...   
232129  http://steamcommunity.com/profiles/76561198209...   

                           posted item_id  \
0        Posted November 5, 2011.    1250   
1           Posted June 24, 2014.  251610   
2              Posted February 3.  248820   
3        Posted October 14, 2013.  250320   
4          Posted April 15, 2014.  211420   
...                           ...     ...   
231919    Posted August 15, 2014.     440   
231921     Posted August 2, 2014.  304930   
232047      Posted July 31, 2015.  265630   
232127  Posted December 20, 2015.  304050   
232129         Posted February 3.     570   

                                                helpful  recommend  \
0                                        No ratings yet       True   
1       15 of 20 people (75%) found this review helpful       True   
2                                        No ratings yet       True   
3        2 of 2 people (100%) found this review helpful       True   
4       35 of 43 people (81%) found this review helpful       True   
...                                                 ...        ...   
231919                                   No ratings yet       True   
231921                                   No ratings yet       True   
232047                                   No ratings yet       True   
232127                                   No ratings yet       True   
232129     0 of 2 people (0%) found this review helpful      False   

                                                   review  
0       Simple yet with great replayability. In my opi...  
1       I know what you think when you see this title ...  
2       A suitably punishing roguelike platformer.  Wi...  
3       This game... is so fun. The fight sequences ha...  
4                                                 Git gud  
...                                                   ...  
231919  TF2 is alot of fun and its really good but the...  
231921                              Fun game with friends  
232047                                        So Fun!! :D  
232127  This game is great. The only thing is,Why cant...  
232129  DOTA 2.normal skill - 2ez for me>high skill - ...  

[59036 rows x 7 columns]

Ademas eliminaremos los registros nulos de reviews(0.05%) dado que al momento de hacer analisis de sentimiento podria ser cualquier opcion (Negative, Neutral o Positive), al no ser un numero tan signficativo los eliminaremos

In [50]:
data_normalizado = data_normalizado.dropna(subset=['review'])
data_normalizado

user_id  \
0           76561197970982479   
1                     js41637   
2                   evcentric   
3                       doctr   
4                   maplemage   
...                       ...   
231919     SKELETRONPRIMEISOP   
231921      76561198141079508   
232047            ShadowYT100   
232127  bestcustomurlevermade   
232129      76561198209894493   

                                                 user_url  \
0       http://steamcommunity.com/profiles/76561197970...   
1                    http://steamcommunity.com/id/js41637   
2                  http://steamcommunity.com/id/evcentric   
3                      http://steamcommunity.com/id/doctr   
4                  http://steamcommunity.com/id/maplemage   
...                                                   ...   
231919    http://steamcommunity.com/id/SKELETRONPRIMEISOP   
231921  http://steamcommunity.com/profiles/76561198141...   
232047           http://steamcommunity.com/id/ShadowYT100   
232127  http://steamcommunity.com/id/bestcustomurlever...   
232129  http://steamcommunity.com/profiles/76561198209...   

                           posted item_id  \
0        Posted November 5, 2011.    1250   
1           Posted June 24, 2014.  251610   
2              Posted February 3.  248820   
3        Posted October 14, 2013.  250320   
4          Posted April 15, 2014.  211420   
...                           ...     ...   
231919    Posted August 15, 2014.     440   
231921     Posted August 2, 2014.  304930   
232047      Posted July 31, 2015.  265630   
232127  Posted December 20, 2015.  304050   
232129         Posted February 3.     570   

                                                helpful  recommend  \
0                                        No ratings yet       True   
1       15 of 20 people (75%) found this review helpful       True   
2                                        No ratings yet       True   
3        2 of 2 people (100%) found this review helpful       True   
4       35 of 43 people (81%) found this review helpful       True   
...                                                 ...        ...   
231919                                   No ratings yet       True   
231921                                   No ratings yet       True   
232047                                   No ratings yet       True   
232127                                   No ratings yet       True   
232129     0 of 2 people (0%) found this review helpful      False   

                                                   review  
0       Simple yet with great replayability. In my opi...  
1       I know what you think when you see this title ...  
2       A suitably punishing roguelike platformer.  Wi...  
3       This game... is so fun. The fight sequences ha...  
4                                                 Git gud  
...                                                   ...  
231919  TF2 is alot of fun and its really good but the...  
231921                              Fun game with friends  
232047                                        So Fun!! :D  
232127  This game is great. The only thing is,Why cant...  
232129  DOTA 2.normal skill - 2ez for me>high skill - ...  

[59006 rows x 7 columns]

In [51]:
Herr.analizar_datos(data_normalizado)

Nombre Tipos de Datos Únicos  % de Valores No Nulos  % de Valores Nulos  \
0    user_id       [<class 'str'>]                  100.0                 0.0   
1   user_url       [<class 'str'>]                  100.0                 0.0   
2     posted       [<class 'str'>]                  100.0                 0.0   
3    item_id       [<class 'str'>]                  100.0                 0.0   
4    helpful       [<class 'str'>]                  100.0                 0.0   
5  recommend      [<class 'bool'>]                  100.0                 0.0   
6     review       [<class 'str'>]                  100.0                 0.0   

   Cantidad de Valores Nulos  
0                          0  
1                          0  
2                          0  
3                          0  
4                          0  
5                          0  
6                          0

Ahora necesitamos que la columna posted, guarde solo el año ya que es necesario este valor para luego ser consumido desde la api

In [52]:
data_normalizado['posted'] = data_normalizado['posted'].apply(Herr.extraccion_anio)
data_normalizado

user_id  \
0           76561197970982479   
1                     js41637   
2                   evcentric   
3                       doctr   
4                   maplemage   
...                       ...   
231919     SKELETRONPRIMEISOP   
231921      76561198141079508   
232047            ShadowYT100   
232127  bestcustomurlevermade   
232129      76561198209894493   

                                                 user_url posted item_id  \
0       http://steamcommunity.com/profiles/76561197970...   2011    1250   
1                    http://steamcommunity.com/id/js41637   2014  251610   
2                  http://steamcommunity.com/id/evcentric   None  248820   
3                      http://steamcommunity.com/id/doctr   2013  250320   
4                  http://steamcommunity.com/id/maplemage   2014  211420   
...                                                   ...    ...     ...   
231919    http://steamcommunity.com/id/SKELETRONPRIMEISOP   2014     440   
231921  http://steamcommunity.com/profiles/76561198141...   2014  304930   
232047           http://steamcommunity.com/id/ShadowYT100   2015  265630   
232127  http://steamcommunity.com/id/bestcustomurlever...   2015  304050   
232129  http://steamcommunity.com/profiles/76561198209...   None     570   

                                                helpful  recommend  \
0                                        No ratings yet       True   
1       15 of 20 people (75%) found this review helpful       True   
2                                        No ratings yet       True   
3        2 of 2 people (100%) found this review helpful       True   
4       35 of 43 people (81%) found this review helpful       True   
...                                                 ...        ...   
231919                                   No ratings yet       True   
231921                                   No ratings yet       True   
232047                                   No ratings yet       True   
232127                                   No ratings yet       True   
232129     0 of 2 people (0%) found this review helpful      False   

                                                   review  
0       Simple yet with great replayability. In my opi...  
1       I know what you think when you see this title ...  
2       A suitably punishing roguelike platformer.  Wi...  
3       This game... is so fun. The fight sequences ha...  
4                                                 Git gud  
...                                                   ...  
231919  TF2 is alot of fun and its really good but the...  
231921                              Fun game with friends  
232047                                        So Fun!! :D  
232127  This game is great. The only thing is,Why cant...  
232129  DOTA 2.normal skill - 2ez for me>high skill - ...  

[59006 rows x 7 columns]

In [53]:
nones = data_normalizado['posted'].isna()
contador = nones.sum()
contador

10097

In [54]:
data_normalizado['posted'] = pd.to_numeric(data_normalizado['posted'],errors='coerce')
mediana = np.nanmedian(data_normalizado['posted'])
data_normalizado['posted'] = data_normalizado['posted'].fillna(mediana)

In [55]:
data_normalizado['posted'] = data_normalizado['posted'].astype(int)

In [56]:
data_normalizado = data_normalizado.drop('helpful',axis=1)

In [57]:
Herr.analizar_datos(data_normalizado)

Nombre Tipos de Datos Únicos  % de Valores No Nulos  % de Valores Nulos  \
0    user_id       [<class 'str'>]                  100.0                 0.0   
1   user_url       [<class 'str'>]                  100.0                 0.0   
2     posted       [<class 'int'>]                  100.0                 0.0   
3    item_id       [<class 'str'>]                  100.0                 0.0   
4  recommend      [<class 'bool'>]                  100.0                 0.0   
5     review       [<class 'str'>]                  100.0                 0.0   

   Cantidad de Valores Nulos  
0                          0  
1                          0  
2                          0  
3                          0  
4                          0  
5                          0

### Carga de datos

In [58]:
ruta_archivo = '../datasets/australian_reviews.csv'

data_normalizado.to_csv(ruta_archivo,index=False,encoding='utf-8')
